packages/babel-plugin-transform-react-constant-elements/src/index.js

```diff
@@ -11,15 +11,38 @@ export default function () {
         return;
       }
 
+      // Elements with refs are not safe to hoist.
       if (path.isJSXIdentifier({ name: "ref" }) && path.parentPath.isJSXAttribute({ name: path.node })) {
         return stop();
       }
 
+      // Ignore identifiers & JSX expressions.
       if (path.isJSXIdentifier() || path.isIdentifier() || path.isJSXMemberExpression()) {
         return;
       }
 
-      if (!path.isImmutable()) stop();
+      if (!path.isImmutable()) {
+        // If it's not immutable, it may still be a pure expression, such as string concatenation.
+        // It is still safe to hoist that, so long as its result is immutable.
+        // If not, it is not safe to replace as mutable values (like objects) could be mutated after render.
+        // https://github.com/facebook/react/issues/3226
+        if (path.isPure()) {
+          const expressionResult = path.evaluate();
+          if (expressionResult.confident) {
+            // We know the result; check its mutability.
+            const resultNode = t.valueToNode(expressionResult.value);
```

Comment:
I just realized `valueToNode` throws if it is given a type that can't be converted to an AST Node, I'm not entirely sure I'm willing to assume that `isPure()` will cover all those cases. It likely does, but since they are not related from a code standpoint it makes me hesitant.

Thoughts?